### DISCUSSION TITLE
Opening files with valid multi-language grammar throws errors

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1533] [Q&A]
Langium version: `3.0.0`
Package name: `langium`

## Steps To Reproduce

1. Create a multi-language example
2. Register languages that has the same file extension
3. Override `ServiceRegistry` and create keywords for those languages to switch between them

Link to code example:

## The current behavior

If I opened a file that has a with a valid language grammar (see language id), it throws errors:

![image](https://github.com/eclipse-langium/langium/assets/87622592/4738838c-3447-4ea0-a055-864f7944f1c6)

But if I delete the content and then add the content again, it works just fine (no need to delete the language keyword, which is here `info`):

![image](https://github.com/eclipse-langium/langium/assets/87622592/936b495f-e191-41ce-910e-4e2a9e80e9d7)

This only happens at the start of running the extension; it works fine for that specific file only after the modification:

![image](https://github.com/eclipse-langium/langium/assets/87622592/e3ce7a05-c3a7-4bb0-b092-97e8d02d8f9e)

## The expected behavior

No errors.

**Comment by cdietrich** [comment]
How did you implement 3

**Reply by Yokozuna59** [reply]
The current approach I'm taking is something like this:

```ts
export class NewServiceRegistry implements ServiceRegistry {
  private documents: () => LangiumDocuments;
  protected services: Record<string, LangiumServices> = {};

  public constructor(shared: LangiumSharedServices) {
    this.documents = () => shared.workspace.LangiumDocuments;
  }

  public get all(): readonly LangiumServices[] {
    return Object.values(this.services);
  }

  public register(language: LangiumServices): void {
    this.services[language.LanguageMetaData.languageId] = language;
  }

  public getServices(uri: URI): LangiumServices {
    const type = this.detectServiceType(uri);
    return this.services[type];
  }

  private detectServiceType(uri: URI): string {
    if (uri.fragment && uri.fragment !== 'default') {
      return uri.fragment;
    } else {
      const content = this.documents().all.find((e) => e.uri.toString() === uri.toString());
      const text: string = content?.textDocument.getText() ?? '';
      try {
        return detectType(text);
      } catch (error) {
        return 'default';
      }
    }
  }
}
```

**Comment by msujew** [comment]
I've converted this into a discussion, since this isn't a Langium bug.

IIRC your service registry looks at the documents service and attempts to discern what kind of language is within the document. However, during the document loading stage (in the [WorkspaceManager](https://github.com/eclipse-langium/langium/blob/abc61f5bcf08b806bc6a5810dd5bf961a47f6865/packages/langium/src/workspace/workspace-manager.ts#L157)), the document to-be-loaded isn't placed in the documents list yet, as it hasn't been created. See [this code](https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/workspace/documents.ts#L424-L425).

You probably need to adjust some of the logic of the document factory [here](https://github.com/eclipse-langium/langium/blob/abc61f5bcf08b806bc6a5810dd5bf961a47f6865/packages/langium/src/workspace/documents.ts#L176-L179) to figure out the service type based on the content of the file without looking into the `LangiumDocuments` service.
